website/docs/guides/migration-guides/migrating-from-v2-to-v3.md

```diff
@@ -0,0 +1,193 @@
+
+Before, the number of workers was determined statically by configuration at the time of application startup. With KafkaFlow v3, a new method was introduced that supports calculating the number of available workers periodically. This allows the application to adapt constantly and also being able to scale the number of workers during the application's lifetime.
+
+To accommodate this requirement multiple changes were made to the core of KafkaFlow, next we will go through some of these changes.
+
+### 4.1 Dynamic Worker Pool Scaling Configuration ([#429](https://github.com/Farfetch/kafkaflow/issues/429))
+
+Two new overrides were added to the consumer configuration. Nevertheless, the previous method of statically determining the worker count is still available.
+
+The new overrides added were:
+
+- `WithWorkersCount(Func<WorkersCountContext, IDependencyResolver, Task<int>> calculator)`
+- `WithWorkersCount(Func<WorkersCountContext, IDependencyResolver, Task<int>> calculator, TimeSpan evaluationInterval)`
+
+The difference between both overrides is that the first works with a predefined frequency of 5 minutes between calls to the delegate that returns the number of workers. While in the second override that frequency can be user-defined.
+
+Additionally, the delegate receives a `WorkersCountContext` that sends some consumer-related context to the callback like the Consumer Name, the Consumer Group Key and the Consumer Partitions Assignment. Also, the `IDependencyResolver` is made available in case some dependencies need to be resolved to determine the worker count.
+
+This means that a consumer worker count can be configured like this:
+
+```csharp
+.AddConsumer(
+    consumerBuilder => consumerBuilder
+        .Topic("test-topic")
+        .WithGroupId("group1")
+        .WithWorkersCount((context, resolver) => {
+          // Use whatever logic to determine the worker count
+          return GetWorkerCount(context, resolver);
+        }, TimeSpan.FromMinutes(15))
+)
+```
+
+### 4.2 Improved Mechanism for Signalling Message Processing Completion ([#426](https://github.com/Farfetch/kafkaflow/issues/426))
+
+For the KafkaFlow engine to adapt to the dynamic change of workers, the way the messages are signaled as completed also needed to be updated because the workers can only be scaled after all the messages in the worker's buffers are signaled as completed.
+
+O achieve this a new method in the message context was provided to signal the message completion `.Complete()`. Additionally, a new property called `AutoMessageCompletion` was added to signal if KafkaFlow should signal the message as completed automatically which is the default behavior. This property is useful for scenarios where we don't want to complete the message right away, for example in Batch Consume scenarios.
```

Comment:
Does KafkaFlow have any sample in which this can be seen?

Reply:
I don't believe so.

Reply:
Could be a good addition for the future, what do you think?

Reply:
This is probably a very niche scenario where users will need to check the message completion.

They would probably be more interested in using events to understand when the message was completed.

I did tough about not putting this change in the migration guide, but although this change is more on the kafkaflow engine level, there could users that are interested in this change.

But I'm not sure it is worth a sample addressing this. Maybe we can evaluate from the community feedback, if we see some questions (issues, etc...) related to this, maybe we can address that.